daphne_worker/src/dap.rs

```diff
@@ -64,32 +65,117 @@ pub(crate) fn dap_response_to_worker(resp: DapResponse) -> Result<Response> {
-        if self.hpke_receiver_config_list.is_empty() {
-            return None;
-        }
+    async fn get_hpke_config_for(
+        &self,
+        _task_id: &Id,
+    ) -> std::result::Result<Option<HpkeConfig>, DapError> {
+        let kv_store = self.kv("DAP_HPKE_RECEIVER_CONFIG_STORE").map_err(dap_err)?;
+        let keys = kv_store
+            .list()
+            .execute()
+            .await
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+
+        if keys.keys.is_empty() {
+            // Generate a new random HPKE config ID.
+            let hpke_config_id: u8 = rand::thread_rng().gen();
+            let name = format!("hpke_receiver_config/{}", hpke_config_id);
+
+            // Generate a new HPKE receiver config.
+            let hpke_receiver_config =
+                HpkeReceiverConfig::gen(hpke_config_id, self.global_config.hpke_kem_id);
 
-        // Always advertise the first HPKE config in the list.
-        Some(&(self.hpke_receiver_config_list[0]).config)
+            // Store newly generated HPKE receiver config into KV.
+            kv_store
+                .put(&name, hpke_receiver_config.clone())
+                .map_err(|e| DapError::Fatal(e.to_string()))?
+                .execute()
+                .await
+                .map_err(|e| DapError::Fatal(e.to_string()))?;
+
+            // Cache the new HPKE receiver config into DaphneWorkerConfig.
+            let mut hpke_receiver_config_mutex_guard =
+                self.hpke_receiver_config.lock().map_err(|e| {
+                    DapError::Fatal(format!(
+                        "Failed to lock hpke_receiver_config: {}",
+                        e.to_string()
+                    ))
+                })?;
+            let hpke_receiver_config_map = hpke_receiver_config_mutex_guard.deref_mut();
+            hpke_receiver_config_map.insert(hpke_config_id, hpke_receiver_config.clone());
+
+            // Add HPKE receiver config in list.
+            Ok(Some(hpke_receiver_config.config))
+        } else {
+            // Return the first HPKE receiver config.
+            let builder = kv_store.get(&keys.keys[0].name);
+            if let Some(hpke_receiver_config) = builder
+                .json::<HpkeReceiverConfig>()
+                .await
+                .map_err(|e| DapError::Fatal(e.to_string()))?
+            {
+                Ok(Some(hpke_receiver_config.config))
+            } else {
+                Ok(None)
+            }
+        }
     }
 
-    fn can_hpke_decrypt(&self, _task_id: &Id, config_id: u8) -> bool {
-        self.get_hpke_receiver_config_for(config_id).is_some()
+    // This is called by the Leader when handling the `/upload` request.
+    // Get HPKE receiver config from KV.
+    async fn can_hpke_decrypt(
+        &self,
+        _task_id: &Id,
+        config_id: u8,
+    ) -> std::result::Result<bool, DapError> {
+        Ok(self
+            .get_hpke_receiver_config_for(config_id)
```

Comment:
Let's check that cache here before hitting KV. just like we do for hpke_decrypt(). Hint: Consider having get_hpke_receiver_config_for() check the cache and populate on a miss. That way you don't have to duplicate this code in multiple places.

Reply:
Great suggestion! This actually helped me to fix what Clippy was complaining about.